extra features to the input proposal (#1953)
dotnet/Silk.NET: documentation/proposals/Proposal - Multi-Backend Input.md

## Custom List Types

These are relatively simple list wrappers with the events fired when state changes.
All of them have one same event called `InputEvent`. This event is fired for any state changed in the device.

```cs
public partial class Mice : IReadOnlyList<IMouse>
{
    public MouseClickConfiguration ClickConfiguration { get; set; }

>     public event Action<dynamic>? InputEvent;

Author: all the events are structs, that are faster but does not allow inheritance so i can't make a common parent to type it. For me is a really important option to can receive any input and can filtrate it as a type so i don't want to just delete it for now. do you know some way to solve his problem?

Unlike 1.0 and 2.0, this proposal uses `readonly record struct`s as their only argument for the event action. This allows us to provide more information to the event handlers without breaking in the future. These types are farily simple:

```cs

> public readonly record struct ConnectionEvent(IInputDevice Device, DateTime Timestamp, bool IsConnected);

Reviewer: These should be DateTimeOffset and should be nullable and the last parameter of these records. This is to ensure compatibility with backends that do not provide the timestamp at which the event occurred, and also means that this won't need to go through working group approval at all given that it is a modification the structs are explicitly designed for (see the already-approved comment about "This allows us to provide more information to the event handlers without breaking in the future.")

Author: we can't provide an timestamp anyway? Maybe it could be wrong for some milliseconds or less, but in the case of don't receive a timestamp, it's not possible to send a DateTime.Now()?

Reviewer 2: > we can't provide an timestamp anyway? Maybe it could be wrong for some milliseconds or less, but in the case of don't receive a timestamp, it's not possible to send a DateTime.Now()?

If the backend doesnt provide timestamps for inputs, i'd rather us not lie about it doing so, if the user needs a timestamp for an event, they can do `Timestamp ?? DateTime.Now()` themselves

Author: True, thx

public readonly record struct JoystickDownEvent(IJoystick Joystick, DateTime Timestamp, JoystickButton Button);
public readonly record struct JoystickUpEvent(IJoystick Joystick, DateTime Timestamp, JoystickButton Button);
public readonly record struct JoystickHatMoveEvent(IJoystick Joystick, DateTime Timestamp, Vector2 Value, Vector2 Delta);
public readonly record struct JoystickAxisMoveEvent(IJoystick Joystick, DateTime Timestamp, int Axis, float Value, float Delta);
public readonly record struct GamepadDownEvent(IGamepad Gamepad, DateTime Timestamp, JoystickButton Button);
public readonly record struct GamepadUpEvent(IGamepad Gamepad, DateTime Timestamp, JoystickButton Button);
public readonly record struct GamepadThumbstickMoveEvent(IJoystick Joystick, DateTime Timestamp, Vector2 Value, Vector2 Delta);
public readonly record struct GamepadTriggerMoveEvent(IJoystick Joystick, DateTime Timestamp, int Axis, float Value, float Delta);
```

This is the part of this proposal that incorporates the ideas in Enhanced Input Events, and is why this proposal supersedes that one.

> 

Reviewer: This is important background information, please do not remove.

Author: i know, but i think this is really out of place because it's on the middle of the reading. You see, i don't deleted all the information and there's the credits of the Enhanced Input Events proposal in the start of the file
You think we can just move it or something? Part of my proposal is bellow and it can cause misunderstanding

Reviewer: The Working Group disagrees given they have approved this information in its place. I don’t think this is what will cause misunderstanding.

Author: If you're saying, ok

`Triggers` contains the two triggers on thsi gamepad. The values within this list range from 0 to 1: 0 being unpressed, and 1 being fully pressed.

Note the use of the `DualReadOnlyList` type. This is basically just:
```cs
public readonly struct DualReadOnlyList<T> : IReadOnlyList<T>
{
    public readonly T Left;
    public readonly T Right;
}
```

> 

Reviewer: Add this back please.

Author: it was already presented, why you want to i put it if you already explained the DualReadOnlyList<T> implementation because it would appear in the rest of the text?

Reviewer: No idea what you’re saying, but I’m not intentionally removing an API declaration that is used throughout the rest of the proposal. It creates needless confusion.

Author: What I'm trying to say is you already put the implementation of the structure, a little above of where you said to i put back the info of where starts the part referencing the other proposal that I can't remember the name (sry) and i think make it two times is unnecessary

Reviewer: Right, but this isn't an implementation detail, this is an API definition. It is normative and should be part of the specification.

`HandleHatMove` must be called when any value of `JoystickState.Hats` changes.

> # Input Maps

Reviewer: This should be a separate proposal given that there's no interest in having this as a core part of input, rather an extension to the existing input library. This also means we don't need to implement it for 3.0, rather we can do so at anytime in 3.X.

## Presentation

> Input maps are designed to be easyer to the user to handle actions with more than one input.

Reviewer: ```suggestion
Input maps are designed to be easier to the user to handle actions with more than one input.
```

> At first view it can looks useless and a bad practice, but this is complemtelly useful for cross-plataform programs, being games an example.

Reviewer: This is a useless comment and contains unfounded claims you have not backed up.
- Why does it look useless?
- Why does it look like bad practice?
- Why specifically is it useful, rather than just saying it's useful?
- Why is it useful for games specifically?
- Why is it useful for cross-platform programs specifically?

Author: idk, ask for beyley, she said it or something of the kind lol

Reviewer 2: What i had said "looks useless and bad practice" was the combination of multiple events into a single callback, i dont see a reason to combine keyboard and mouse events into a single callback, when you'd have to `if(ev is KeyboardEvent kbEvent)`, im not against input *maps* being in Silk.NET (_as an extension_, **not** core)

what i dont like is `public event Action<dynamic>? InputEvent;`

Author: Oh, reality? Sry so, my fault. I will fix it

Author: > So to clarify, you’re making a proposal you don’t personally understand the usefulness in?

I just don't see anything useless on it, i just misunderstood Beyley's comment. But anyway you have to say, it really looks a little useless at first look lol

> At first view it can looks useless and a bad practice, but this is complemtelly useful for cross-plataform programs, being games an example.

Reviewer: ```suggestion
At first view it can looks useless and a bad practice, but this is completely useful for cross-platform programs, being games an example.
```

> This is an pseudocode example of what the Input Map system solves:

Reviewer: Problems are best described by words. You haven't stated anywhere in this proposal what this actually achieves and why this is a valuable proposal for the working group to consider.

Reviewer: No because it isn't clear what problem you're solving. e.g. what is an input action and why is it useful to hide the actual callbacks behind one?

Author: Oh, sure sure! I really have not seen that i should follow that way when i wrote but now i can see it, thx

Without inputMaps:
```cs
inputContext.Update();

private Dictionary<string, bool> pressedKeys = new();

inputContext.Gamepads.ThumbstickMove += @event => {Move(@event.Value)};
inputContext.Joysticks.AxisMove += @event => {Move(@event.Value)};

private Move(Vector2D<float> value)
{
    // Code for movement here
}
```

With inputMaps:
```cs
inputContext.Update();

// This single line get rid of the two in the example above. TWO ISN'T THE LIMIT!
inputContext.Map.GetAction("Move").Axis += @event => {Move(@event.Value)};

private Move(Vector2D<float> value)
{
    // Code for movement here
}
```

> ## Implementation

Reviewer: Consider changing this header, implementation details should be INFORMATIVE TEXT only rather than a normative part of the proposal. I know that's probably not what you intend this section to specify, which is why I say change the header.

Author: yeah, the section don't teach you to implement nothing, just maps the base of the system... i will fix

> Each InputContext have one reference to an InputMap.

Reviewer: Why have you done the ownership this way around? Surely it makes sense for an input map to own an input context and not the other way around given that the map is essentially an extension/grouping of the existing functionality? This also relates to my other comment about making this a separate proposal.

Author: the map is an extension/grouping of the receiving of input events. It's on the same level of a ButtonDown or a ButtonUp, the only difference is because it will filtrate and operate the inputs before send it back.
i really can't see how a InputMap can own an InputContext because of it

Reviewer: Nonetheless, we are eager to ensure that this is strictly an extension to the core API and not part of the core API itself, so consider other ways to express this relationship. Inheritance maybe? Composition?

Author: I really can't see how i can make it work with inheritance becuase the inputContext are really complex and there's a lot od useless information to the maps, so i will try to do composition, also because you will always get an InputContext from the core window so (i really don't want to make it using extension methods, i believe in composition supremacy)

Reviewer: For reasons such as this, I really wish you had waited to make this proposal to be honest, this isn't a one person job by any stretch of the imagination and we're likely to have to back-track on anything decided here once we actually start implementing it, which could be years from now.

Feel free to come up with an alternative design, but I don't think we can have enough confidence having so little available information. Both the windowing and input proposals came from multiple maintainers experimenting with potential designs and actively collaborating in a short space of time to come up with what feels the most natural both to use and implement, and we just don't have this information available for your proposal because it is a one person job. That's not to say that one person proposals aren't accepted, they definitely have been in the past, but historically these have been the proposals we've come to regret implementing the most. And we're just not interested in doing that experimentation right now because we have so many more pressing priorities before we can even look at extensions such as this. This is also one of the many reasons I put a comment to make this a separate proposal, so that it doesn't impact our much higher priority work.

Author: I understand, i will put this section in a separated proposal and try to improve the maximum that i can

Author: Also you want to hold it for a while until someone help to develop it better? Posting the proposal looks like the best way people can see and modify it, but it will be a problem if it gets rejected

Reviewer: I think placing it in a separate proposal and merging it with an `(On Hold)` prefix is the best course of action.

> This is how the InputMap class work:

Reviewer: An API declaration does not show _how_ something works.

Author: Sure, bad word use

```cs
public partial class InputMap
{

>     public InputMapAction GetAction(string name);

Reviewer: Consider providing documentation comments and/or MoSCoW/RFC-style requirements as done throughout the 3.0 proposals. As it stands I have no idea what any of this must do and would have no idea how to implement it from this document alone.

Author: I really want to use more comments but i was a little afraid because I was only using this proposal as a reference. I will check how it is done in another one and try to describe better

    public void AddAction(InputMapAction action, string name);
    public void RemoveAction(string name);

>     public static InputMap Load(string path);

Reviewer: File format? Should this be included in the name?

Author: Yes the extension of the file is .File Format

Jokes apart, I don't have a specific file format yet, but it should be a problem for the system that should just read and parse its content. So while I don't have creativity to think of something to use as an standard you can have a inputMap.FileFormat 👍

Author: Should not*** be a problem

>     public static InputMap Load(string path);

Reviewer: Do we want to make this exclusive to system IO? Why not provide a ROSpan overload?

Reviewer 2: imo it shouldnt be silk's job to do any IO here, the user should just supply a string for the map contents, and "Save" should just return a string for the saved map

Author: Really? I was thinking about just returning the serialized content as a string but for me if the library does the IO work, better

If it's not silk's job i will change so, it's not like it's a problem

    public static void Save(InputMap source, string path);

>     public static string Compile(InputMap source);

Reviewer: What does this even mean?

Author: If you're talking about the `compile` method, it serializes and returns the result as a string. I know, the name is bad. I will change it, i really forgot the word for a moment

}
```

> `GetAction` method is the main entry point of the system. It allows the use to get an mapped action by it configurated name.

Reviewer: ```suggestion
`GetAction` method is the main entry point of the system. It allows the use to get a mapped action by its configured name.
```

```cs

> public partial class InputMapAction

Reviewer: Generally I think more definition needs to be done about what's in scope and what's out of scope for this model.

Author: This model is really raw, huh... I need to improve it a lot

{

>     public bool IsPressed { get; }

Reviewer: What does this mean for actions that aren't boolean on/off actions?

Author: Good question... I was thinking about creating classes for each type of input, you think it's better in that way? As it's a class there's no reason to put all possible returns in the same type

Reviewer: I do not think it's better in that way. I don't know what could make it better to be honest.

Author: Tbh i really do not want to do it with separate classes also... I will try to think a way to everything works just inside the InputMap class, without unnecessary data access

>     public event Action<AxisInputEvent> Axis;

Reviewer: How does this work with mapping WASD as a substitute for a thumbstick?

Author: It's on the input map syntax example.

If an input doesn't return a vector, you can write

```
- silkVectorClass [

]
```

And the specified inputs inside it, -1 +1, for each axis in the vector class, one input per line.
So like (pseudocode):

```
- Vector2D [
    - A // -x
    - D // +x
    - S // -y
    - W  // +y
]
```

If it's not explained enough in the proposal pls notify

Reviewer: Okay, this is not explained in the proposal (at all) particularly your latter code snippet.

Author: I will try to improve it more. Also improve the syntax explanation with some other features like ranges and dead radius (this is the name? If it's not sry i will obviously search when i can)

Also some way of returning more customizable values or use modifiers like shift or ctrl

}
```

## Configuration and saving

> The input map system's save system is one of the most important to it working. It helps to make the configuration easyer as faster.

Reviewer: Please use a spellchecker (e.g. Grammarly)
```suggestion
The input map system's save system is one of the most important to it working. It helps to make the configuration easier and faster.
```

Author: Yeah, i have it installed on my work pc but not in my personal one lol i will do it (also back to Duolingo or a better English course probably)

> Instead of needing to hardcode all the actions, the user should create a file with a special syntax. This is an example of the sintax of the map system's saved data:

Reviewer: ```suggestion
Instead of needing to hardcode all the actions, the user should create a file with a special syntax. This is an example of the syntax of the map system's saved data:
```

> Instead of needing to hardcode all the actions, the user should create a file with a special syntax. This is an example of the sintax of the map system's saved data:

Reviewer: Implying that this is the only route of entry?

Author: These are the two options and the file configuration should be the best choice, i should considerate another option or I'm misunderstanding something?

```
"move_up":
    - KeyName.Up
    - JoystickButton.DPadUp

"move_down":
- KeyName.Down

> - JoystickButton.DPadDown

Reviewer: Are the padding differences significant? What are the parsing rules here?

Author: 905 & 906 are exactly to show padding isn't a rule, but i will write about the optional use of the white space in this case

> "move":

Reviewer: How does this handle multiple controllers e.g. for multiple players?

Author: Thing that i don't think again (or at least tried but can't) lol

In the case of the same input device, different bindings

In the case of different devices, It could return the data of each device separated (like a get mappedByDevice()) but it can broke a lot in some cases...

(Sry busy now i will add more data)

    - Vector2D [
        -KeyName.Down
        -KeyName.Up
        -KeyName.Left
        -KeyName.Right

>     ]

Reviewer: What is the significant of Vector2D and why does the contained elements have more than 2 elements? Are you intending for this to be the WASD mapping I alluded to earlier? In which case, what are the rules for this? e.g. how do we map each key to a specific value on the axes.

Author: Vector2D is to define the specified vector. It can be any vector, 2, 3 or 4

The question of why there's 4 keys is because the first is for the negative value and the second is for the positive value, being it for each axis

So the size of the list will aways be the axis count * 2

And yes, is what you asked before

>     - Joystick.Axis

Reviewer: Which joystick will this pick up on? Any?

Author: I need to say, I don't understand so much what devices "joystick" and "gamepad" are referring. The idea should be any, but now I'm thinking about the fact that a ps4 controller have 2 thumbsticks and i forgot it completely lol

Can you explain me a little what exactly these devices represent? Like i really never head something called anything for these names (obviously in some cases but pretty obvious ones like the atari controller or the one that some flying machines uses, that's i think is out of where silk acts lol), just everything as controller

Also like what input will be returned by these weird car controllers or joysticks (are these joysticks that it refers?)?

```

> The main syntax may somewhat resemble YAML, but it's much more symple and easy to parse.

Reviewer: Why are we making our own file format? This is scope creep and beyond our engineering resources to execute. Using a well-supported format such as JSON is perfectly fine, a custom format is over engineering.

Reviewer 2: i personally would lean towards ini due to its simplicity for non-developers to read/write.
Although the question is, are end users (eg. a player of a game built on Silk.NET) ever intended to modify this file? Or is this mapping file for developers only? The API being hardcoded to read/write to a path implies to me that its meant to be handled by end-users aswell, and JSON can be quite daunting to newcomers

Author: I don't want to use existing file formats like json or YAML (that is a base for syntax) because an entire parser for these ones will be unnecessarily expensive for something so simple. Also i want a file format easy for anyone to edit, being developers or users. Also i want something easy and fast to read and parse

This file format can use just a single loop to iterate per line and the state machine-based structure makes you and the machine easily understand the scopes

So i think it can be more optimized and cheaper if we do our own file format that does JUST what we want with an easy to parse syntax instead of use a complex data structure that needs a lot of operations to parse

Reviewer: I can guarantee that System.Text.Json will be infinitely faster than any parser that the Silk.NET team has the ability to maintain itself. The same is likely true for extant YAML parsers.

This proposal creates unnecessary maintenance effort for the Silk.NET team that the team won't want to be dealing with, and will likely refuse to implement or merge on those grounds.

Author: understand. I will search about the YAML and JSON parsers and if you're right in this case, i will change to one of these two

> The main syntax may somewhat resemble YAML, but it's much more symple and easy to parse.

Reviewer: ```suggestion
The main syntax may somewhat resemble YAML, but it's much more simple and easy to parse.
```

> The main structure is based on a list, with it main feature having the line separation as the main way to separate the information. Also, the parsing is based on an finite-state model. The last data seted will recive the next configurations until a higher level information be setted.

Reviewer: You can't say it's based on a finite state machine without specifying the states of said state machine! How are we supposed to implement that if so? Also, surely this is an implementation detail that should be non-normative?

> The main structure is based on a list, with it main feature having the line separation as the main way to separate the information. Also, the parsing is based on an finite-state model. The last data seted will recive the next configurations until a higher level information be setted.

Reviewer: Grammar/spelling issues. I have no idea what the last part of this sentence is trying to say so I haven't attempted to reword.

Author: Looks right for me, but i will try to rewrite the explanation of how the state machine works

Is it possible to create graphics and diagrams with github's markdown?

> The label are wrote inside quotes ("") and ended with a collom (:). This starts a new `InputMapAction`.

Reviewer: Consider providing a grammar for this? e.g. ANTLR

Author: Lol no sry, it was late and i was tired, i correct it

Reviewer: What?

Author: i misunderstood the message sry. if i continue with this file type i will do one

> The label are wrote inside quotes ("") and ended with a collom (:). This starts a new `InputMapAction`.

Reviewer: ```suggestion
The label is written inside of quotes ("") and delimited by a colon (:). This starts a new `InputMapAction`.
```

> each new input are declared with a minnus sign (-) and a stace. After that, the name of the correct enum and the correct input.

Reviewer: ```suggestion
each new input are declared with a minus sign (-) and a space. After that, the name of the correct enum and the correct input.
```

> each new input are declared with a minnus sign (-) and a stace. After that, the name of the correct enum and the correct input.

Reviewer: But how do we discriminate between inputs from different devices or different types of devices?

Author: The input map system can't discriminate between input device types, the objective of it is exactly that

About the different device discrimination, i will work on it in the next commit

> Input system can easly return vectors. after the minus sign, it's possible choose a vector type to recive (`Vector2D`, `Vector3D` or `Vector4D`) and the inputs that will math witch each axis, respectively.

Reviewer: "return vectors" the example above uses keyboard buttons as well and does not specify in terms of requirements for how to map this into a vector. How is the axis matching done? (i.e. mathematically)

Author: Yeah, this parsing needs to be a lot more structured

> Input system can easly return vectors. after the minus sign, it's possible choose a vector type to recive (`Vector2D`, `Vector3D` or `Vector4D`) and the inputs that will math witch each axis, respectively.

Reviewer: ```suggestion
Input system can easily return vectors. After the minus sign, it's possible choose a vector type to receive (`Vector2D`, `Vector3D` or `Vector4D`) and the inputs that will match which each axis, respectively.
```

> Comments are not allowed in these files.

Reviewer: This is such an arbitrary restriction and is one of the millions of reasons why I am reluctant to have a custom format

Author: I really can't see why put comments on it, it's just a structure to save data. Json also doesn't allow comments as standard, but should not be a problem to implement it, so i will try to add some rules sbout

Author: i want to discuss a little more about it, can we do it on discord?

Reviewer: No, here is fine.

> To Load an `InputMap` from a file, the static method `Load` should be used. in any case of edition during runtime, it's possible to save the changes using the method `Save`.

Reviewer: ```suggestion
To load an `InputMap` from a file, the static method `Load` should be used. If the map is edited at runtime, it's possible to save the changes using the method `Save`.
```
